File: Library/src/main/java/org/endeavourhealth/library/model/search/DownloadByQueryOptions.java
package org.endeavourhealth.library.model.search;

import lombok.Getter;
import org.endeavourhealth.library.model.imq.ECLQueryRequest;
import org.endeavourhealth.library.model.requests.QueryRequest;

@Getter
public class DownloadByQueryOptions {
  private QueryRequest queryRequest;
  private ECLQueryRequest eclSearchRequest;
  private int totalCount;
  private String format;
  private boolean includeDefinition;
  private boolean includeCore;
  private boolean includeLegacy;
  private boolean includeSubsets;
  private boolean subsetsOnOwnRow;
  private boolean im1id;

  public DownloadByQueryOptions setQueryRequest(QueryRequest queryRequest) {
    this.queryRequest = queryRequest;
    return this;
  }

  public DownloadByQueryOptions setEclSearchRequest(ECLQueryRequest eclSearchRequest) {
    this.eclSearchRequest = eclSearchRequest;
    return this;
  }

  public DownloadByQueryOptions setTotalCount(int totalCount) {
    this.totalCount = totalCount;
    return this;
  }

  public DownloadByQueryOptions setFormat(String format) {
    this.format = format;
    return this;
  }

  public DownloadByQueryOptions setIncludeDefinition(boolean includeDefinition) {
    this.includeDefinition = includeDefinition;
    return this;
  }

  public DownloadByQueryOptions setIncludeCore(boolean includeCore) {
    this.includeCore = includeCore;
    return this;
  }

  public DownloadByQueryOptions setIncludeLegacy(boolean includeLegacy) {
    this.includeLegacy = includeLegacy;
    return this;
  }

  public DownloadByQueryOptions setIncludeSubsets(boolean includeSubsets) {
    this.includeSubsets = includeSubsets;
    return this;
  }

  public DownloadByQueryOptions setSubsetsOnOwnRow(boolean subsetsOnOwnRow) {
    this.subsetsOnOwnRow = subsetsOnOwnRow;
    return this;
  }

  public DownloadByQueryOptions setIm1id(boolean im1id) {
    this.im1id = im1id;
    return this;
  }
}

File: src/models/DownloadByQueryOptions.ts
import z from "zod";

import { ECLQueryRequestSchema } from "./ECLQueryRequest";
import { QueryRequestSchema } from "./QueryRequest";

// export interface DownloadByQueryOptions {
//   queryRequest?: QueryRequest;
//   eclSearchRequest?: ECLQueryRequest;
//   totalCount?: number;
//   format?: string;
//   includeDefinition?: boolean;
//   includeCore?: boolean;
//   includeLegacy?: boolean;
//   includeSubsets?: boolean;
//   subsetsOnOwnRow?: boolean;
//   im1id?: boolean;
// }

export const DownloadByQueryOptionsSchema = z.strictObject({
  queryRequest: QueryRequestSchema.optional(),
  eclSearchRequest: ECLQueryRequestSchema.optional(),
  totalCount: z.number().optional(),
  format: z.string().optional(),

  includeDefinition: z.boolean().default(false),
  includeCore: z.boolean().default(false),
  includeLegacy: z.boolean().default(false),
  includeSubsets: z.boolean().default(false),
  subsetsOnOwnRow: z.boolean().default(false),
  im1id: z.boolean().default(false)
});

export type DownloadByQueryOptions = z.output<typeof DownloadByQueryOptionsSchema>;

export function isDownloadByQueryOptions(value: unknown): value is DownloadByQueryOptions {
  return DownloadByQueryOptionsSchema.safeParse(value).success;
}
